doc/advanced-testing-mini-course/README.md

```diff
@@ -234,32 +189,61 @@ assertProgram msg prog = do
 
-3.  Collect timing information about how long each command takes to execute on average.
+    -   Stateless (or pure) property-based testing tries to *approximate* proof by induction in math. For example the following is the proposition that addition is associative for integers, *forall i j k. (i + j) + k == i + (j + k)*. It looks almost exactly like the property you'd write in a property-based test, but of course this test passing isn't a proof of the proposition, still a step in the right direction if we want to be serious about program correctness.
 
-See also
---------
+    -   XXX: Stateful property-based testing using state machines, like we seen in this lecture, tries to approximate proof by structural induction on the sequence of inputs. Or inductive invarint method?!
 
--   Why state machines over other forms of specifications? E.g. unit test-suite.
     -   Executable (as the REPL exercise shows, but also more on this later)
+
+    -   Same state machine specification can be used for concurrent testing (Lec 2)
+
     -   Mental model
-    -   Gurevich's generalisation of the Church-Turing thesis
+
     -   Already heavily used in distributed systems (later we'll see how the model becomes the implementation)
 
+## Excerises
+
+0.  If you're not comfortable with Haskell, port the above code to your favorite programming language.
+
+1.  Add a `Reset` `Command` which resets the counter to its initial value.
+
+2.  Implement shrinking for programs.
+
+3.  Write a REPL for the state machine. Start with the initial state, prompt the user for a command, apply the provided command to the step function and display the response as well as the new state, rinse and repeat.
+
+    (For a SUT as simple as a counter this doesn't make much sense, but when the SUT get more complicated it might make sense to develope the state machine specification first, demo it using something like a REPL or some other simple UI before even starting to implement the real thing.)
+
+4.  Collect timing information about how long each command takes to execute on average.
+
+## See also
+
+-   The original QuickCheck [paper](https://dl.acm.org/doi/pdf/10.1145/357766.351266) by Koen Claessen and John Hughes (2000) that introduced property-based testing in Haskell.
+
+-   John Hughes' Midlands Graduate School 2019 [course](http://www.cse.chalmers.se/~rjmh/MGS2019/) on property-based testing, which covers the basics of state machine modelling and testing. It also contains a minimal implementation of a state machine testing library built on top of Haskell's QuickCheck;
+
+-   Lamport's [Computation and State Machines](https://www.microsoft.com/en-us/research/publication/computation-state-machines/)
+
+    (2008) 
+
+-   "Can one generalize Turing machines so that any algorithm, never mind how ab- stract, can be modeled by a generalized machine very closely and faithfully?"
```

Comment:
```suggestion
-   "Can one generalize Turing machines so that any algorithm, never mind how abstract, can be modeled by a generalized machine very closely and faithfully?"
```